### PR TITLE
url_store_state implementation

### PR DESCRIPTION
**Pull request by Echo11ovo**
# Description

Implemented URL-based data transfer, fixing the local storage issue.

# Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

- [x] This change requires a documentation update

# How Has This Been Tested?

-  Test 1: When a new user enters the page, there is no data inside the page's local storage, and the page is rendered by using default values
image

-  Test 2: Input values are saved even if the user switches models during use.
image
image

-  Test3: By sharing a URL with data, the page is able to render the corresponding value more in line with the URL
image
image

# Remaining issues

- Data can't go to local storage when switching units, still exploring related solution ideas
- There seems to be a problem with the bottom two user inputs via dropdown boxes that continually triggers a page refresh, rendering them unusable

### CODE REVIEW

File: pages/home.py [hunk 1]
```diff
@@ -106,20 +113,49 @@ def update_store_inputs(
...
+def update_inputs(selected_model, units_selection, url_search, stored_data):
     if selected_model is None:
         return no_update
+
     units = UnitSystem.IP.value if units_selection else UnitSystem.SI.value
-    return input_environmental_personal(selected_model, units)
+
+    # Parse URL parameters
+    url_params = parse_qs(url_search.lstrip("?"))
+    url_params = {k: v[0] if len(v) == 1 else v for k, v in url_params.items()}
```
Comment (FedericoTartarini) [inline pages/home.py:142]: this code is very hard to read

File: pages/home.py [hunk 2]
```diff
@@ -106,20 +113,49 @@ def update_store_inputs(
...
+    print(f"url_params: {url_data}")
+
+    return inputs, url_search
 
 
 @callback(
     Output(ElementsIDs.INPUT_SECTION.value, "children"),
     Input(ElementsIDs.MODEL_SELECTION.value, "value"),
     Input(ElementsIDs.UNIT_TOGGLE.value, "checked"),
+    Input("url", "search"),
```
Comment (FedericoTartarini) [inline pages/home.py:129]: do not use a string "url" but instead create an attribute to ElementsIDs called url

**Review by FedericoTartarini — COMMENTED** [review]
It is hard for me to review the code since I am not familiar with it. Is it working? Have you tested it extensively?

File: pages/home.py (hunk 1 above)
Comment (Echo11ovo) [reply pages/home.py:142]: This line of code is a dictionary derivation that iterates through all the key-value pairs in url_params, extracting the value of the key if it corresponds to a list with only one element to make the change, or making the change to the entire list if the list of values contains more than one element, such as when we follow up with a single set that has more than one value in compare

**Comment by Echo11ovo** [comment]
The part about URL storage data has been fully tested, but at this stage there are still two problems, the first one is that `Metabolic Rate (met)` and `Clothing Level (clo)` will keep refreshing when inputting and cannot write the value, it seems to be a problem with the `dmc component`, and the second one is that the value will not be properly stored in local storage when **switching units**，we are still thinking about the solution.

File: pages/home.py (hunk 2 above)
Comment (Echo11ovo) [reply pages/home.py]: fit it

Comment (Echo11ovo) [reply pages/home.py]: I will fix it

File: pages/home.py [hunk 3]
```diff
@@ -75,14 +79,17 @@
 
 @callback(
     Output(MyStores.input_data.value, "data"),
+    Output("url", "search", allow_duplicate=True),
```
Comment (FedericoTartarini) [inline pages/home.py:82]: this also should be from ElementsId an not a string

**Review by FedericoTartarini — APPROVED** [review]

**Comment by FedericoTartarini** [comment]
@Echo11ovo please fix the merging issue that is now there. 

I am also very worried about these issues. Are they breaking the application? You need to fix them before we should integrate the Pull reuqest.

* Data can't go to local storage when switching units, still exploring related solution ideas
* There seems to be a problem with the bottom two user inputs via dropdown boxes that continually triggers a page refresh, rendering them unusable

File: pages/home.py (hunk 3 above)
Comment (Ha02hen) [reply pages/home.py]: This will be resolved.

**Comment by FedericoTartarini** [comment]
@Echo11ovo I really like the code you implemented in the SMA application, it is simple to read and maintain. I noticed that here you have so many more functions and code changes. Can we re-write the code so it is much more simple? For example, why do we need `store-state.py` and `url-store-state.py`

**Comment by Echo11ovo** [comment]
> @Echo11ovo I really like the code you implemented in the SMA application, it is simple to read and maintain. I noticed that here you have so many more functions and code changes. Can we re-write the code so it is much more simple? For example, why do we need `store-state.py` and `url-store-state.py`

Actually, I didn't use `store_state.py` and `url_store_state.py` because they were based on our original understanding of the project, which first implemented data persistence and then `url` mapping, but when deployed to the cbe comfort project, we found the `app.py->home .py->input_environment_personal.py` This structure is difficult to implement for data transfer. So in this pull request, I didn't use these two code files. I will remove them. The implementation is now basically the same as the SMA application, except that there are more changes to the data interface, such as `model selection`, `user input`, and some part of the  `calculate`, which is why so many changes are needed.

**Review by FedericoTartarini — APPROVED** [review]

**Pull request merged**